rclcpp_components/CMakeLists.txt

```diff
@@ -61,9 +61,29 @@ ament_target_dependencies(component_container_mt
   "rclcpp"
 )
 
+add_executable(
+  component_container_isolated
+  src/component_container_isolated.cpp
+)
+target_link_libraries(component_container_isolated component_manager)
+ament_target_dependencies(component_container_isolated
+  "rclcpp"
+)
+
+add_executable(
+  component_container_mt_isolated
```

Comment:
Though adding this is a couple lines of code, I think it's better to not add it to avoid confusion.

My reasons are:
- The more options, the more confusing which one is the best.
- The multithreaded executor doesn't execute anything in parallel, except if multiple nodes were added to it or if there're a node with more than one callback group (most nodes are just using the default callback group only).
  In this case, we would be launching one executor per node, which means that if the node is only using the default callback group the `MultithreadedExecutor` will perform worse than the single threaded one (multiple threads doing completely mutually exclusive work).
- Even in the case of nodes having multiple callback groups, creating more than one multithreaded executor will end up in more threads that cores, which usually doesn't perform well either.

Reply:
There's a fourth reason:

- The multithreaded executor is only taking advantage of the multiple threads in `spin()`, and not in the other methods like `spin_until_future_complete()`.

Though that's something we should improve ...

Reply:
>  a node with more than one callback group

There are definitely real examples of more complex nodes that do have multiple callback groups that could make use of this multi-threaded executor container where each can have an isolated executor. It's the same rationale for why we want the single threaded version of this -- we want them to be in the same process but be `isolated` from each other to prevent unintended consequences of callback groups that were not designed to operate at the same time, operating at the same time. But an application that can leverage this container type is designed for internally multiple callback groups being executed so it makes perfect sense. 

Some some simplistic application, sure, but isn't the point of ROS 2 to be supporting real, deployed, complex applications that ROS 1 could not? This is part of that, IMO. 

I do not believe this is the point of too much confusion. I'd be confused as a user if the multi-threaded version of this didn't exist - because it would make me ask "why not when there's a non-isolated `mt` version?". 

The only reason I think that we should not have an `mt-isolated` container is if we depreciate the normal `mt` container as well so there is no `mt` component containers at all.

Reply:
> It's the same rationale for why we want the single threaded version of this -- we want them to be in the same process but be isolated from each other to prevent unintended consequences of callback groups that were not designed to operate at the same time, operating at the same time.

Sorry, I haven't been able to follow everything here, but I was skimming and saw this, and it stood out to me.

If two callbacks were designed to not be run at the same time, then they should be in the same mutually-exclusive callback group. If they're in different nodes, then something really bad has gone wrong in the design. Anything else and they could be run at the same time.

A node that by design requires a dedicated single-threaded executor to address this kind of issue is, in my opinion, a bad solution. If that is a requirement, then the node should instead create a thread and have its own single-threaded executor for those sensitive callbacks. That way it is not negatively impacted by which executor it is added to in a container. The downside to that approach is scaling, having every node have its own thread and executor will not scale as well, but used judiciously it's fine.

Now if the issue is just congestion, like two callbacks from two different nodes can in principle share a multi-threaded executor, but in practice one starves the callbacks of the other, then this approach can be fine I think. So I'm not opposed to having this as an option.

To reduce confusion (and again I haven't reviewed the changes so I don't know how feasible this is), it might be better to simply have CLI options (or ros param options) for the single container to control which executors it supports.

Reply:
>  If they're in different nodes, then something really bad has gone wrong in the design. 

I don't disagree. That was probably not a good illustration. The congestion example you provide is a much better example of why one might want to isolate components with different multi-threaded executors -- so that one doesn't hog the entire pool. 

> To reduce confusion (and again I haven't reviewed the changes so I don't know how feasible this is), it might be better to simply have CLI options (or ros param options) for the single container to control which executors it supports.

I'd be fine with that. Then, shouldn't we also apply that same option to the normal component container for consistency and remove the existing `mt`?

Reply:
> Sure... but that's clearly subject to later improvements - but my argument above still applies: ROS 2 is built for real commercial applications that are going to be complex.

My proposal of not having a "multithreaded" option in the isolated container is because it's a really unusual approach in any real application.
The most usual approaches are:

- Single threaded event loop. This might be combined with a thread pool of workers to handle blocking tasks (usually, the event loop has some way to interact with tasks in the thread pool to make the combination nice). e.g. python asyncio runtime
- One single threaded event loop per core. This might also be combined with a thread pool of workers to handle blocking tasks. e.g. boost asio can be configured this way, some rust runtimes, etc
- One multithreaded event loop. This might be combined with a thread pool of workers as well. e.g. boost::asio can be configured this way, default behavior of most rust runtimes, etc.

The "single threaded executor" is kind of equivalent to a single threaded event loop.
The "multi threaded executor" is kind of equivalent to a multi threaded event loop.
The "isolated executor per node" is kind of equivalent to having one event loop per core. This approach doesn't scale well if you have many nodes (because you will have much more threads than cores).

The "isolated multithreaded executor per node" is not equivalent to anything you usually see in event loop runtimes.
It really doesn't scale well, as you create as many threads as cores per each multithreaded executor (by default), and you  easily end up having a lot of threads.

---

It's only 5 extra lines of codes and the discussion doesn't seem to be making much progress, so I will accept this as-is.